daffodil-core/src/test/scala/org/apache/daffodil/dsom/walker/TestWalker.scala

```diff
@@ -0,0 +1,335 @@
+   * DFDL choices or if they are embedded in another element (which corresponds to being in a Record).
+   * In the end, if they are not already inside a Record, then they are put there. Choices cannot
+   * have fields, only possible sub-types, so anything that would be a "field" has a wrapper Record
+   * put around it.
+   * @param schemaNode the node to convert to a Choice data type
+   * @return a Choice data type as described above
+   */
+  private def choiceNodeToChoiceSchema(schemaNode: TestNode): DataType = {
+    val childList: ListBuffer[DataType] = new ListBuffer()
+    for (child <- schemaNode.children) {
+      if (child.name.isEmpty) {
+        childList += schemaNodeToRecordType(child)
+      } else {
+        // In the case where the child has a name, we embed it in an additional TestSchema,
+        // since we must add some DataType rather than a Field.
+        val fieldList: ListBuffer[TestField] = ListBuffer()
+        fieldList += nodeToField(child)
+        val choiceOption: TestSchema = newRecordSchema(child, fieldList.toList)
+        childList += RecordDataType(choiceOption)
+      }
+    }
+    ChoiceDataType(childList.toList)
+  }
+
+  /**
+   * Local helper method to appropriately convert a TestNode into an appropriate Record
+   * Data Type.  Records and Choices get routed to other helper methods, and Arrays are handled in the method.
+   * @param schemaNode the node to convert to a Record Data Type
+   * @return the finalized Record Data type
+   */
+  private def schemaNodeToDataType(schemaNode: TestNode): DataType = {
+    schemaNode.fieldType match {
+      case ArrayType =>
+        if (schemaNode.isSimple) {
+          ArrayDataType(new DataType(getSimpleType(schemaNode)))
+        } else if (schemaNode.children.nonEmpty) {
+          ArrayDataType(schemaNodeToRecordType(schemaNode))
+        } else {
+          throw new IllegalStateException(
+            s"Array Node $schemaNode either needs to be simple or have child nodes!"
+          )
+        }
+      case RecordType => schemaNodeToRecordType(schemaNode)
+      case ChoiceType => choiceNodeToChoiceSchema(schemaNode)
+      case _ => schemaNode.simpleType.get
+    }
+  }
+
+  /**
+   * In general, this method can be used to extract the TestSchema node from the tuple result
+   * of schemaNodeToRecordSchema.  However, it is really only used to kick off the conversion from
+   * the root TestNode.
+   * @param schemaNode the node from which a TestSchema will be extracted
+   * @return the processed TestSchema represented by the input
+   */
+  private def getRecordSchema(schemaNode: TestNode): TestSchema = {
+    schemaNodeToRecordType(schemaNode).childSchema
+  }
+
+}
+
+/**
+ * The sole purpose of this Companion Object is to contain this one static variable.
+ */
+object RecordWalker {
+  /**
+   * Used to suppress all of the extra printing from running test cases
+   */
+  final val PRODUCTION_MODE: Boolean = true
```

Comment:
I'm concerend that this test code is fairly complex, and that it's pretty easy for there to be a small bug in here which could lead to us not realizing there's an issue with the walker. I'm wondering if there's a less complex way to test the DSOM walker to verify that it's doing the right thing?

We really just want to validate that the DSOM walker creates the correct events in the correct order. We don't really need to use these events to build anything to test that it's walking correctly. Perhaps it would be simpler if the TestDSOMWalker just received an ordered list of expected events and properties of those events. As the TestDSOMWalker walks the tree and receives events, it just ensures that they match the expected list of events.

There's probably multiple approaches, but the first thing that comes to my mind is to implement customer the walker.Views that have a custom comparision function to compare things we want to test. Note that we don't necessarily need to check all properties, for example, it might make sense to ignore group, groupMembers, simpleType, complexType, etc. since these are all implied correct if we get the correct events. So one of these might look look something like this:
```scala
case class TestRootView(
  override isSimpleType: Boolean,
  override isComplexType: Boolean,
  ...)
  extends walker.RootView {
  
  override def equals(that: AnyRef): Boolean =
    that match {
      case rv: RootView =>
        this.isSimpleType == rv.isSimpletype &&
        this.isComplexType == rv.isComplexType &&
        ...
    case _ => false
  }
}
```
We'd have a concrete TestView for each walker View we could get from the walker. Then the TestDSOMWalker could look something like this:
```scala
class TestDSOMWalker(expectedEvents: Seq[TestView]) {
  var events = expectedEvents
  def onSomeEvent(view: walker.SomeView): Unit = {
    val head :: tail = events
    assertTrue(head == view)
    events = tail 
  }
}
```
Then to test, this becomes:
```
def testWalker1() = {
  val expectedEvents = Seq(
    TestRootView(....),
    TestComplexTypeView(...),
    TestSequenceView(...),
    ...
  )
  ...
  val testWalker = new TestDSOMWalker(expectedEvents)
  testWalker.walk(pf.rootView)
}
```

There's still a little complexity here--we need to create these concrete views and ensure we're testing the right properties, including start/end events. But I think it's a little easier to verify that the walker implementation is doing the right thing with a simpler approach.

Reply:
I think I wound up with a solution that is similar to this idea; I had a testing file generate a Stack-like structure based on different View types it received, and the Tests were based around specific expected events in that List.  Since I was rewriting the Tests, I chose to automate as little as possible, I only automated the Simple Types since checking that they work was the same way for each one.  However, for each of the Complex types, I wanted to check different things like the name in some cases.  This is all in the latest revision.

Reply:
I did not wind up ignoring things like complexType and simpleType in the testComplexTypes() test case, but I do have a flag for that in the new testing class if that is desired - let me know if you prefer that.